File: 84/solution_stack.cpp
// monotonic stack

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> s;
        heights.push_back(0);
        int n = heights.size();
        int result = 0, i = 0;
        
        while(i < n) {
            if(s.empty() || heights[i] >= heights[s.top()]) {
                s.push(i);
                i++;
            } else {
                int h = heights[s.top()]; s.pop();
                int w = s.empty() ? i : i - s.top() - 1;
                result = max(result, h * w);
            }
        }
        
        return result;
    }
};

File: 84/solution_stack.py
# monotonic stack

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        result = 0
        s = []
        heights.append(0)
        n = len(heights)
        i = 0
        
        while i < n:
            if not s or heights[i] >= heights[s[-1]]:
                s.append(i)
                i += 1
            else:
                h = heights[s[-1]]; s.pop()
                w = i if not s else i - s[-1] - 1
                result = max(result, h * w)
        
        return result
